File: frontend/demo/component/grid/grid-column-freezing.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/button';
import '@vaadin/grid';
import { columnBodyRenderer } from '@vaadin/grid/lit.js';
import { getPeople } from 'Frontend/demo/domain/DataService';
import type Person from 'Frontend/generated/com/vaadin/demo/domain/Person';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('grid-column-freezing')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  private items: Person[] | undefined;

  protected override async firstUpdated() {
    const { people } = await getPeople();
    this.items = people;
  }

  protected override render() {
    return html`
      <vaadin-grid .items="${this.items}">
        <!-- tag::snippet1[] -->
        <vaadin-grid-column
          frozen
          header="Name"
          auto-width
          flex-grow="0"
          ${columnBodyRenderer<Person>(
            (person) => html`${person.firstName} ${person.lastName}`,
            []
          )}
        ></vaadin-grid-column>
        <!-- end::snippet1[] -->
        <vaadin-grid-column path="email" auto-width></vaadin-grid-column>
        <vaadin-grid-column path="address.phone" auto-width></vaadin-grid-column>
        <vaadin-grid-column path="profession" auto-width></vaadin-grid-column>
        <vaadin-grid-column path="address.street" auto-width></vaadin-grid-column>
        <!-- tag::snippet2[] -->
        <vaadin-grid-column
          frozen-to-end
          auto-width
          flex-grow="0"
          ${columnBodyRenderer(
            () => html`<vaadin-button theme="tertiary-inline">Edit</vaadin-button>`,
            []
          )}
        ></vaadin-grid-column>
        <!-- end::snippet2[] -->
      </vaadin-grid>
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/grid/GridColumnFreezing.java
package com.vaadin.demo.component.grid;

import java.util.List;

import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.data.renderer.LitRenderer;
import com.vaadin.flow.data.renderer.Renderer;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;

@Route("grid-column-freezing")
public class GridColumnFreezing extends Div {

    public GridColumnFreezing() {
        Grid<Person> grid = new Grid<>(Person.class, false);
        // tag::snippet1[]
        grid.addColumn(createPersonRenderer()).setHeader("Name").setFrozen(true)
                .setAutoWidth(true).setFlexGrow(0);
        // end::snippet1[]
        grid.addColumn(Person::getEmail).setHeader("Email").setAutoWidth(true);
        grid.addColumn(person -> person.getAddress().getPhone())
                .setHeader("Phone").setAutoWidth(true);
        grid.addColumn(Person::getProfession).setHeader("Profession")
                .setAutoWidth(true);
        grid.addColumn(person -> person.getAddress().getStreet())
                .setHeader("Street").setAutoWidth(true);
        // tag::snippet2[]
        grid.addColumn(createActionRenderer()).setFrozenToEnd(true)
                .setAutoWidth(true).setFlexGrow(0);
        // end::snippet2[]

        List<Person> people = DataService.getPeople();
        grid.setItems(people);

        add(grid);
    }

    private static Renderer<Person> createPersonRenderer() {
        return LitRenderer.<Person> of("${item.fullName}")
                .withProperty("fullName", Person::getFullName);
    }

    private static Renderer<Person> createActionRenderer() {
        return LitRenderer.<Person> of(
                "<vaadin-button theme=\"tertiary-inline\">Edit</vaadin-button>");
    }

    public static class Exporter // hidden-source-line
            extends DemoExporter<GridColumnFreezing> { // hidden-source-line
    } // hidden-source-line
}
